Enforce FPE instead of NaN + some fixes for FPE

shahor02 opened the pull request:
@pzhristov @davidrohr  @ktf  this PR is a potential can or worms, following https://alice.its.cern.ch/jira/browse/O2-3754?focusedId=305287&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-305287 with the help of Peter I've enforced the FPE where before NaN was produced and already found 2 cases just in the FST.

We have to decide if we want to merge / deploy it. Perhaps, to avoid surprises in the online processing, one can make the activation of the FPE trap conditional to some env. var. 

Then there is a comment from Peter that libXML>=20913 triggers FPE. At the moment I am not activating the detection of the libXML version since in the O2 we are using 20903 (but ubuntu 22.04 system version is 20913, mostly because I am not sure how to detect the libxml/xmlversion.h path for `target_include_directories` for https://github.com/AliceO2Group/AliceO2/compare/dev...shahor02:pr_FPE?expand=1#diff-5ff8176836cd9c83c8422cc102ee8f78d80757498bbce94e6ce5ef0a3e54d12eR95. Any suggestion?

shahor02 committed:
Trap FEs instead of producing NaN, particular FE trapping can be overridden.

By default FE_INVALID|FE_DIVBYZERO|FE_OVERFLOW ( i.e. 8|4|1 = 13) is trapped, one can
override this set by defining env.var with bits corresponding to FEs
to disable. E.g. O2_DISABLE_FPE_TRAP=13 will disable all FEs, while
O2_DISABLE_FPE_TRAP=9 will leave only FE_DIVBYZERO enabled.

shahor02 committed:
Fix for FE_OVERFLOW in PHOS

2 problems: (1)if the chi2 is too large, its implicit conversion to float leads to overflow
(2) in my FST run the getChi2 was called before it was assigned in the fit, and since it was
not initialized in the fitter c-tor, it returned a random huge value. Initialization helps
to avoid random value but one should avoid querying getChi2 w/o fit.

shahor02 commented:
@peressounko @kharlov as I wrote in https://github.com/AliceO2Group/AliceO2/pull/11804/commits/f08367d5a2b7b37ea84c735384f107c156fad692, in the FST the getChi2 was queried w/o being assigned in the fit, could you please check?

shahor02 committed:
Fix FPE in rANS: avoid taking log of 0

davidrohr commented:
Well, in general I think having this optuonal to check for FPEs makes sense. I have an optional.setting for my GPU code which I regulalry check for FPEs.

The problem to activate it in general with 3rd party code is that then you can get FPE that you cannot fix from everywhere in the 3rd party code. I have also seen some from ROOT and some other libraries, and also e.g. when using the cuda profiler one gets FPEs. Not sure if we want to enable it by default, but in any case we would then still have the option to disable it. Don't have a strong opinion.

Kind Regards
David Rohr

Sent from my mobile. (Excuse the typos!)

On 22 August 2023 14:50:57 CEST, Ruben Shahoyan ***@***.***> wrote:
>@pzhristov @davidrohr  @ktf  this PR is a potential can or worms, following https://alice.its.cern.ch/jira/browse/O2-3754?focusedId=305287&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-305287 with the help of Peter I&#39;ve enforced the FPE where before NaN was produced and already found 2 cases just in the FST.
>
>We have to decide if we want to merge / deploy it. Perhaps, to avoid surprises in the online processing, one can make the activation of the FPE trap conditional to some env. var. 
>
>Then there is a comment from Peter that libXML&gt;=20913 triggers FPE. At the moment I am not activating the detection of the libXML version since in the O2 we are using 20903 (but ubuntu 22.04 system version is 20913, mostly because I am not sure how to detect the libxml/xmlversion.h path for `target_include_directories` for https://github.com/AliceO2Group/AliceO2/compare/dev...shahor02:pr_FPE?expand=1#diff-5ff8176836cd9c83c8422cc102ee8f78d80757498bbce94e6ce5ef0a3e54d12eR95. Any suggestion?
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/AliceO2Group/AliceO2/pull/11804
>
>-- Commit Summary --
>
>  * Enforce FPE instead of NaN
>  * Fix FPE in rANS: avoid taking log of 0
>  * Fix for FE_OVERFLOW in PHOS
>
>-- File Changes --
>
>    A Common/Utils/src/fpu.c (25)
>    M Detectors/PHOS/reconstruction/include/PHOSReconstruction/CaloRawFitter.h (4)
>    M Detectors/PHOS/reconstruction/src/AltroDecoder.cxx (4)
>    M Utilities/rANS/src/FrequencyTable.cxx (3)
>    M cmake/O2AddExecutable.cmake (5)
>
>-- Patch Links --
>
>https://github.com/AliceO2Group/AliceO2/pull/11804.patch
>https://github.com/AliceO2Group/AliceO2/pull/11804.diff
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/AliceO2Group/AliceO2/pull/11804
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***&gt;

shahor02 commented:
@davidrohr  the problem is that we have much more potential sources of FPE on CPU code, will be difficult to check them with dedicated tests. At the moment I've modified the trapping commit as 
```
    Trap FEs instead of producing NaN, particular FE trapping can be overridden.
    
    By default FE_INVALID|FE_DIVBYZERO|FE_OVERFLOW ( i.e. 8|4|1 = 13) is trapped, one can
    override this set by defining env.var with bits corresponding to FEs
    to disable. E.g. O2_DISABLE_FPE_TRAP=13 will disable all FEs, while
    O2_DISABLE_FPE_TRAP=9 will leave only FE_DIVBYZERO enabled.
```
@chiarazampolli, in principle, I would port this to async tag to detect as many as possible FPEs.

alibuild commented:
Error while checking build/O2/fullCI for 9e467d10be13c48b71497cfbf94319354836e2b8 at 2023-08-22 19:02:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/11804/9e467d10be13c48b71497cfbf94319354836e2b8/build_O2_fullCI/pretty.html).

chiarazampolli commented:
> @davidrohr the problem is that we have much more potential sources of FPE on CPU code, will be difficult to check them with dedicated tests. At the moment I've modified the trapping commit as
> 
> ```
>     Trap FEs instead of producing NaN, particular FE trapping can be overridden.
>     
>     By default FE_INVALID|FE_DIVBYZERO|FE_OVERFLOW ( i.e. 8|4|1 = 13) is trapped, one can
>     override this set by defining env.var with bits corresponding to FEs
>     to disable. E.g. O2_DISABLE_FPE_TRAP=13 will disable all FEs, while
>     O2_DISABLE_FPE_TRAP=9 will leave only FE_DIVBYZERO enabled.
> ```
> 
> @chiarazampolli, in principle, I would port this to async tag to detect as many as possible FPEs.

Hi @shahor02 , once merged, I will create a new tag, sure. 
Thanks!

shahor02 commented:
fullCI fails due to the FPE in the FT0 geometry creation: https://alice.its.cern.ch/jira/browse/O2-4057.

shahor02 committed:
Temporarily change O2_DISABLE_FPE_TRAP to O2_ENABLE_FPE_TRAP

Given that there are potentially many places producing FPE if with FPE trapping enbaled,
in order to merge this PR changing default to NOT enable trapping with possibility
to enable it via env.var O2_ENABLE_FPE_TRAP with bits corresponding to enabled FEs:
FE_INVALID(=8)|FE_DIVBYZERO(=4)|FE_OVERFLOW(=1)

shahor02 commented:
Temporarily change O2_DISABLE_FPE_TRAP to O2_ENABLE_FPE_TRAP

 Given that there are potentially many places producing FPE if with FPE trapping enabled, in order to merge this PR changing default to NOT enable trapping with possibility to enable it via env.var O2_ENABLE_FPE_TRAP with bits corresponding to enabled FEs:
```
    FE_INVALID(=8)|FE_DIVBYZERO(=4)|FE_OVERFLOW(=1)
```

alibuild commented:
Error while checking build/O2/fullCI for 6346f2b3ac7e14f4f0cd25ab4bd6a0a03bdcff0d at 2023-08-23 00:25:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
/sw/SOURCES/O2/11804-slc8_x86-64/0/Common/Utils/src/fpu.cxx:30:10: error: inclusion of deprecated C++ header 'fenv.h'; consider using 'cfenv' instead [modernize-deprecated-headers]
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/11804/6346f2b3ac7e14f4f0cd25ab4bd6a0a03bdcff0d/build_O2_fullCI/pretty.html).

davidrohr commented:
> @davidrohr the problem is that we have much more potential sources of FPE on CPU code, will be difficult to check them with dedicated tests. At the moment I've modified the trapping commit as

Sure, that is what I meant. Once we globally enable it, it can crash randomly in many places, also data driven. But still no bad idea to find such places, perhaps not for online running.

But if we get crashes from 3rd party libraries, which is not unlikely, it is not so easy to solve.

shahor02 commented:
macOS failure is unrelated (@ktf, looks like there is a general problem), merging
